### DISCUSSION TITLE
[Example] Generating a platformer level

### DISCUSSION THREAD

**Discussion by danbraik** [AlexeyBond/godot-constraint-solving#26] [Show and tell]
Hello,
thank you for this wonderful addon. After reading some of the discussion, I understood better how to use Precondition. It gave me the idea to use WFC as a better autotile feature than builtin one.
Moroever, the idea is to draw the shape of a platformer level and let WFC fill it. I had to use a special tile for the AIR since it is a 2D side view.
You can find the prototype project so everyone can learn from it. [test-wfc.zip](https://github.com/user-attachments/files/16581410/test-wfc.zip)

So, the settings is:
- we have a tileset with marked tile for the platform (grass, dirt) and tiles for the air (purple one and foliage decoration)
- we have a sample map to explicitly teach how to connect platform tiles
- we have a negative map to reject wrong combinations
- we have a level tilemaplayer to draw the shape of the level
- we have a special new class `WFC2DPreconditionPlaftorm`

Demo:

Level map:
![image](https://github.com/user-attachments/assets/ae109d93-4d91-4808-b41c-97f933b26e75)

Result in game:
![image](https://github.com/user-attachments/assets/11f03f89-5e18-4b45-b9b2-694328d8d4b5)

Now I am just wondering if it will take more times to set the rules than actually and manually draw levels myself ;-) 
Have a nice day!

**Comment by AlexeyBond** [comment]
Hi.

I'm glad someone besides myself could get how to use this feature 😆 

I did think about making a more generic precondition that would map a tile from target map to certain domain. Mostly for cases when initial level layout is generated by another procedural generator (such as [Gaea](https://github.com/BenjaTK/Gaea), or alike) that isn't wrapped in a precondition and WFC just polishes the details. I guess it would work for your case as well.

I did also consider creating a WFC-based autotiling-like functionality that would work directly in the editor. But that seems to be too complicated and isn't likely to be significantly more useful than the existing autotiling functionality (with terrains).

**Comment by danbraik** [comment]
Hi,
thank you for the answer.
Exactly, my need was to set a domain per tile. Thank for Gaea I didn't know about and seems to go hand in hand with your addon.

Actually, I think the WFC-based autotiling would be better because it seems the builtin autotiling only updates tiles around a placed tile, so it can have a lot of aberration during the draw. 
Moreover, to work in editor, the algorithm should support "incremental update" where the user can place a tile and the algo reacts to that. Would it be complicated to update the algo to a behavior like [Townscaper](https://oskarstalberg.com/Townscaper) ? (and maybe later a tool in editor to draw levels like that https://github.com/mxgmn/WaveFunctionCollapse#constrained-synthesis )

**Reply by AlexeyBond** [reply]
> Moreover, to work in editor, the algorithm should support "incremental update" where the user can place a tile and the algo reacts to that. Would it be complicated to update the algo to a behavior like [Townscaper](https://oskarstalberg.com/Townscaper) ? (and maybe later a tool in editor to draw levels like that https://github.com/mxgmn/WaveFunctionCollapse#constrained-synthesis )

"constrained synthesis" tool should be easy enough to make (except visualization of non-fixed tiles before they are collapsed). In fact, that's what algorithm already does, all what's left is to make it run in editor.

Townscaper-like behavior would be a bit more complicated, but pretty much doable. I did make a [tiny game](https://alexeybond.itch.io/degenerative-dungeon) that uses WFC to re-generate a small sector of a map. It's should possible to replicate a Townscaper mechanic (in 2d) using the same approach, but additionally constraining the central cell to tiles corresponding to user choice and making the algorithm to not change existing tiles when possible. I think, I actually may try to make a demo with such mechanic this weekend.

I think, in both cases the most difficult part will be making a proper UI and editor integration.

**Reply by AlexeyBond** [reply]
Here is what I could make: https://alexeybond.itch.io/dungeonscaper
